README.md
# Refine v3.1
*Reads specific lines of text documents.*
<br/>

The *refine* command was created to read specific lines of text files. This command, usually does the service that *head* and *tail* do. However, it has some extra parameters that are not present in other commands in a natural way.

The interest of this project is to create a command that facilitates the search for specific lines in files, without having to resort to *pipes* or huge codes to obtain the desired output.
<br/><br/>

For example, I want the clean output of line 10:
```sh
$ cat -n file.txt | grep -w "10" | cut -d$'\t' -f2-
```
<br/>

If there is a more efficient way, it will not be different from that. Another example is to display lines 10 to 20:
```sh
$ head -20 file.txt | tail -11
```
<br/>

It is short, but if the file has 15 lines instead of 20, this combination will not do this control and will display 11 lines and not 6, which would be 10 to 15.
<br/><br/>

### Parameters:
<br/>

The *refine* it has the following syntax:
```
$ refn [options] file.txt
```
<br/>

The *refine* it has a series of parameters. See the follow:
```sh
    -f: To read the first lines of the files.
    -l: To read the last lines of the files.
    -d: To read the specific lines of the files.
    -i: To read inverted the specific lines of the files.
    -c: Enable color for f, l, d and i parameters.
    -s: Simple visualization for f, l, d and i parameters.
    -h: To view help information.
```
<br/>

### Comparation and Use:
<br>

The use of *refine* is very simple. Let's start with the basic commands. How was said, this command has the functions of *head* and *tail*, but with some improvements, as can be seen folow:
```sh
$ head -5 LICENSE 
                    GNU GENERAL PUBLIC LICENSE
                       Version 3, 29 June 2007

 Copyright (C) 2007 Free Software Foundation, Inc. <https://fsf.org/>
 Everyone is permitted to copy and distribute verbatim copies
```
```sh
$ refn -f 5 LICENSE 

/home/mauricio/GIT/refine/LICENSE:
         1:                     GNU GENERAL PUBLIC LICENSE
         2:                        Version 3, 29 June 2007
         3: 
         4:  Copyright (C) 2007 Free Software Foundation, Inc. <https://fsf.org/>
         5:  Everyone is permitted to copy and distribute verbatim copies
```
```sh
$ refn -f 5 -s LICENSE 
                    GNU GENERAL PUBLIC LICENSE
                       Version 3, 29 June 2007

 Copyright (C) 2007 Free Software Foundation, Inc. <https://fsf.org/>
 Everyone is permitted to copy and distribute verbatim copies
```
<br/>

Now, compared to the *tail*. As you can see, the last line does not enter the equation if it is empty, just like the *tail*:
```sh
$ tail -5 LICENSE 
into proprietary programs.  If your program is a subroutine library, you
may consider it more useful to permit linking proprietary applications with
the library.  If this is what you want to do, use the GNU Lesser General
Public License instead of this License.  But first, please read
<https://www.gnu.org/licenses/why-not-lgpl.html>.
```
```sh
$ refn -l 5 LICENSE 

/home/mauricio/GIT/refine/LICENSE:
       670: into proprietary programs.  If your program is a subroutine library, you
       671: may consider it more useful to permit linking proprietary applications with
       672: the library.  If this is what you want to do, use the GNU Lesser General
       673: Public License instead of this License.  But first, please read
       674: <https://www.gnu.org/licenses/why-not-lgpl.html>.
```
```sh
$ refn -l 5 -s LICENSE 
into proprietary programs.  If your program is a subroutine library, you
may consider it more useful to permit linking proprietary applications with
the library.  If this is what you want to do, use the GNU Lesser General
Public License instead of this License.  But first, please read
<https://www.gnu.org/licenses/why-not-lgpl.html>.
```
<br/>

Now comes the part that makes sense of the command existance. Let's compare the output of specific numbers of lines, which, in the example, will be listed lines in 26 to 30, as can be seen folow:
```sh
$ head -30 LICENSE | tail -5
want it, that you can change the software or use pieces of it in new
free programs, and that you know you can do these things.

  To protect your rights, we need to prevent others from denying you
these rights or asking you to surrender the rights.  Therefore, you have
```
```sh
$ refn -d 26-30 LICENSE 

/home/mauricio/GIT/refine/LICENSE:
        26: want it, that you can change the software or use pieces of it in new
        27: free programs, and that you know you can do these things.
        28: 
        29:   To protect your rights, we need to prevent others from denying you
        30: these rights or asking you to surrender the rights.  Therefore, you have
```
```sh
$ refn -d 26-30 -s LICENSE 
want it, that you can change the software or use pieces of it in new
free programs, and that you know you can do these things.

  To protect your rights, we need to prevent others from denying you
these rights or asking you to surrender the rights.  Therefore, you have
```
<br/>

Now, to close with golden key, let's analyze the code used to read only a specific line, as in the following example:
```sh
$ cat -n LICENSE | grep -w "  10"
    10    The GNU General Public License is a free, copyleft license for
   446    10. Automatic Licensing of Downstream Recipients.
```
```sh
$ cat -n LICENSE | grep -w "  10" | cut -d$'\t' -f2-
  The GNU General Public License is a free, copyleft license for
  10. Automatic Licensing of Downstream Recipients.
```
```sh
$ refn -d 10 LICENSE 

/home/mauricio/GIT/refine/LICENSE:
        10:   The GNU General Public License is a free, copyleft license for
```
```sh
$ refn -d 10 -s LICENSE 
  The GNU General Public License is a free, copyleft license for
```
<br/>

You want the tenth last line? Here it is:
```sh
$ refn -i 10 /etc/slackpkg/mirrors 

/etc/slackpkg/mirrors:
       353: # https://mirror.slackbuilds.org/slackware/slackware64-current/
```
```sh
$ refn -i 10 -s /etc/slackpkg/mirrors 
# https://mirror.slackbuilds.org/slackware/slackware64-current/
```
<br/>

The result dispenses with comments. So, to conclude, here is a command alternative to facilitate the search for specific lines in files.

        --> NOTE: The commands '-f', '-l', '-d' and '-i' were thought to be used separately.
<br/>

### GNU General Public License:

This repository has scripts that were created to be free software.<br/>
Therefore, they can be distributed and / or modified within the terms of the *GNU General Public License*.

>[General Public License](https://pt.wikipedia.org/wiki/GNU_General_Public_License)
>
>Free Software Foundation (FSF) Inc. 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA

<br/>

### Comments:

In case of bugs, execution problems or packages construction, constructive criticism, among others, please submit a message to one of the contacts below.
<br/><br/>

### Contact:

Autor: Mauricio Ferrari

E-Mail: *m10ferrari1200@gmail.com*

Telegram: *@maurixnovatrento*


ChangeLog
# ChangeLog
All notable changes to this project will be documented in this file.
Todas as mudanças notáveis ​​neste projeto serão documentadas neste arquivo.


## [ 3.1 ] - 2021-09-07 ##################################################

### Added / Adicionado
 - Error message for no file entries.
###

### Changed / Modificado
 - Clean unnecessary codes for '-c' and '-s' parameters.
###

##########################################################################


## [ 3.0 ] - 2021-09-04 ##################################################

### Added / Adicionado
 - Added parameter '-s' for simple view for '-f', '-l', '-d' and '-i'
   parameters.
###

### Changed / Modificado
 - Code refactored and improved using the function getopt_long() for
   parameters selection.
 - Error messages improved.
###

### Removed / Removido
 - Removed parameter 0, replaced by '-s' parameters.
###

##########################################################################


## [ 2.0 ] - 2021-02-14 ##################################################

### Added / Adicionado
 - Added parameter '-i' for read inverted the specific lines of the files.
 - Added parameter 'c' for enable color for '-f', '-l', '-d' and '-i'
   parameters.
###

### Changed / Modificado
 - Changed parameter '-d' for best delimitation.
###

### Removed / Removido
 - Removed parameter '-s', now unnecessary for parameter 'd'.
 - Removed parameter '-a', function now made by the parameter 'd'.
###

### Fixed / Corrigido
 - Fixed exibition of the complete local of the text files.
 - Output adjust of the parameter '0'.
 - Improvements in the selection of parameters.
###

##########################################################################


## [ 1.0 ] - 2021-02-08 ##################################################

### Added / Adicionado
 - Added parameter '-f' for read the first lines of the files.
 - Added parameter '-l' for read the last lines of the files.
 - Added parameter '-a' for read a line from a file.
 - Added parameter '-d' and '-s' for read the specific lines of the files.
 - Added parameter '-h' for view help information.
 - The Parameter '-f0', '-t0', '-a0' and '-d0' has the same result
   as '-f', '-t', '-a' and '-d', but without enumerating the lines.
###

##########################################################################


main.cpp
/**     "refine v3.1" - Reads specific lines of text documents.     **
 **     By Mauricio Ferrari - Date Creation: 2021/02/07.            **
 **                         - Last Update:   2021/09/04.            **/

#include <iostream>
#include <cstring>
#include <string>
#include <vector>
#include <sstream>

#include <getopt.h> /** Include C Library. **/

#include "utils/control_utils.h"
#include "utils/err_msg.h"
#include "utils/file_utils.h"
#include "utils/line_utils.h"
#include "utils/parm.h"
#include "main.h"

using namespace std;

/** Help function. **/

#pragma clang diagnostic push
#pragma ide diagnostic ignored "OCDFAInspection"

void help( char *x ) {
    cout << "refine v3.1 - Reads specific lines of text documents." << endl;
    cout << "\nSyntax: " << x << " [options] text_file" << endl;
    cout << "\nParameters:" << endl;
    cout << "\t-f: To read the first lines of the files." << endl;
    cout << "\t-l: To read the last lines of the files." << endl;
    cout << "\t-d: To read the specific lines of the files." << endl;
    cout << "\t-i: To read inverted the specific lines of the files." << endl;
    cout << "\t-h: To view help information." << endl;
    cout << "\nOptionals:" << endl;
    cout << "\t-c: Enable color for f, l, d and i parameters." << endl;
    cout << "\t-s: Simple visualization for f, l, d and i parameters." << endl;
    exit(0);
}

#pragma clang diagnostic pop

/** Main function. **/

int main( int argc, char **argv ) {

    char *ref_file = basename(argv[0]);

    /** First controls. **/

    if ( argc == 1 ) help(ref_file);
    if ( strcmp(argv[1], "-c") == 0 || strcmp(argv[1], "-s") == 0 ) err_cs_first(ref_file);

    int opt, i;
    int first_line = default_view;
    int last_line = default_view;

    long int delimit_f;
    unsigned long int delimit_l = nullarg;

    char *first_parm = nullptr;
    char *last_parm = nullptr;
    char *delimit_line = nullptr;
    char *invert_line = nullptr;

    char *recover_arg = nullptr;
    char *file;

    bool set_first = false;
    bool set_last = false;
    bool set_delimit = false;
    bool set_invert = false;
    bool set_color = false;
    bool set_simple = false;
    bool discount_arg = false;

    /** Struct for parameters. **/

    const struct option parm[] = {
            {"first",   optional_argument, null_p, parm_f},
            {"last",    optional_argument, null_p, parm_l},
            {"delimit", required_argument, null_p, parm_d},
            {"invert",  required_argument, null_p, parm_i},
            {"color",   no_argument,       null_p, parm_c},
            {"simple",  no_argument,       null_p, parm_s},
            {"help",    no_argument,       null_p, parm_h},
            {null_p,    no_argument,       null_p, null_p},
    };

    /** Parameters selection. **/

    while ( (opt = getopt_long(argc, argv, "f::l::d:i:csh", parm, nullptr)) > 0 ) {

        /** Control for optional parameters. **/

        if ( !optarg && optind < argc && nullptr != argv[optind] && '\0' != argv[optind][0] && '-' != argv[optind][0] ) optarg = argv[optind++];

        switch ( opt ) {

            case 'f': /** Option -f or --first **/

                if ( set_first || set_last || set_delimit || set_invert ) err_parm(ref_file, set_first, parm_f, msg_err_first);

                if ( optarg != nullptr ) {
                    if ( number_validate(optarg) == 0 ) first_line = stoi(optarg, nullptr, 0);
                    else {
                        discount_arg = true;
                        first_parm = optarg;
                    }
                }

                set_first = true;
                break;

            case 'l': /** Option -l or --last **/

                if ( set_first || set_last || set_delimit || set_invert ) err_parm(ref_file, set_last, parm_l, msg_err_last);

                if ( optarg != nullptr ) {
                    if ( number_validate(optarg) == 0 ) last_line = stoi(optarg, nullptr, 0);
                    else {
                        discount_arg = true;
                        last_parm = optarg;
                    }
                }

                set_last = true;
                break;

            case 'd': /** Option -d or --delimit **/

                if ( set_first || set_last || set_delimit || set_invert ) err_parm(ref_file, set_delimit, parm_d, msg_err_delimit);

                delimit_line = optarg;
                set_delimit = true;
                discount_arg = false;
                break;

            case 'i': /** Option -i or --invert **/

                if ( set_first || set_last || set_delimit || set_invert ) err_parm(ref_file, set_invert, parm_i, msg_err_invert);

                invert_line = optarg;
                set_invert = true;
                discount_arg = false;
                break;

            case 'c': /** Option -c or --color **/

                if ( set_color || !set_first && !set_last && !set_delimit && !set_invert ) err_parm(ref_file, set_color, parm_c,nullptr);

                set_color = true;
                discount_arg = true;
                break;

            case 's': /** Option -s or --simple **/

                if ( set_simple || !set_first && !set_last && !set_delimit && !set_invert ) err_parm(ref_file, set_simple, parm_s,nullptr);

                set_simple = true;
                discount_arg = true;
                break;

            case 'h': /** Option -h or --help **/
                help(ref_file);
                break;

            default:
                cerr << ref_file << ": Invalid option or argument not found." << endl;
                exit(1);
        }
    }

    /** Additional controls. **/

    if ( set_color && set_simple ) err_cs(ref_file);
    if ( optind == 1 || argc < 3 ) help(ref_file);

    /** Select delimitation for -f and -l parameters. **/

    if ( set_delimit || set_invert ) {
        int delimit = 0;
        char *test, *option_parm;
        string token;

        if ( set_delimit ) option_parm = delimit_line;
        else option_parm = invert_line;

        if ( check_delimit(option_parm) == 0 ) {
            vector<string> tokens;
            istringstream tokenizer{option_parm};
            while ( getline(tokenizer, token, '-')) tokens.push_back(token);

            for ( const string &data: tokens ) {
                copy(data.begin(), data.end(), token.begin());
                test = &token[0];
                if ( number_validate(test) == 0 ) {
                    if ( delimit == 0 ) delimit_f = stoi(data, nullptr, 0);
                    if ( delimit == 1 ) delimit_l = stoi(data, nullptr, 0);
                } else err_num(ref_file, option_parm, 1);
                delimit++;
            }

        } else {
            if ( number_validate(option_parm) == 0 ) delimit_f = stoi(option_parm, nullptr, 0);
            else err_num(ref_file, option_parm, 0);
        }
    }

    int res = optind;
    if ( discount_arg ) res--;

    /** Error control for -c parameters in the end and for -f/-l parameters without arguments. **/

    for ( i = res; argv[i] != nullptr; i++ ) {
        if ( strcmp(argv[i], "-c") == 0 || strcmp(argv[i], "-s") == 0 ) {
            res++;
            if ( first_parm != nullptr ) recover_arg = first_parm;
            if ( last_parm != nullptr ) recover_arg = last_parm;
        }
    }

    /** Control for optional parameters for -f and -l. **/

    if ( recover_arg != nullptr ) {
        cout << "foi";
        if ( check_files(recover_arg) == 1 ) err_file(ref_file, recover_arg, 0);
        file = realpath(recover_arg, nullptr);
        if ( set_first ) view_lines(set_simple, set_color, parm_f, file, first_line, nullarg);
        if ( set_last ) view_lines(set_simple, set_color, parm_l, file, last_line, nullarg);
    }

    if ( recover_arg == nullptr && argv[res] == nullptr ) err_file(ref_file, nullptr, 1);

    /** Control for entry files. **/

    for ( i = res; argv[i] != nullptr; i++ ) {
        if ( check_files(argv[i]) == 1 ) err_file(ref_file, argv[i], 0);
        file = realpath(argv[i], nullptr);
        if ( set_first ) view_lines(set_simple, set_color, parm_f, file, first_line, nullarg);
        if ( set_last ) view_lines(set_simple, set_color, parm_l, file, last_line, nullarg);
        if ( set_delimit ) view_lines(set_simple, set_color, parm_d, file, delimit_f, delimit_l);
        if ( set_invert ) view_lines(set_simple, set_color, parm_i, file, delimit_f, delimit_l);
    }
    return 0;
}
